File: HexToFile.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <iomanip>

int main() {
    std::string inputFilePath, outputFilePath;
    std::ifstream inputFile;
    std::ofstream outputFile;
    unsigned int hexValue;

    // Prompt user for input file path
    std::cout << "Enter the input hex file path: ";
    std::cin >> inputFilePath;

    // Prompt user for output file path
    std::cout << "Enter the output file path: ";
    std::cin >> outputFilePath;

    // Open the input hex file for reading
    inputFile.open(inputFilePath);
    if (!inputFile.is_open()) {
        std::cerr << "Error opening input file: " << inputFilePath << std::endl;
        return EXIT_FAILURE;
    }

    // Open the output file for writing in binary mode
    outputFile.open(outputFilePath, std::ios::binary);
    if (!outputFile.is_open()) {
        std::cerr << "Error opening output file: " << outputFilePath << std::endl;
        inputFile.close();
        return EXIT_FAILURE;
    }

    // Read hex values from the input file and write binary data to the output file
    while (inputFile >> std::hex >> hexValue) {
        outputFile.put(static_cast<unsigned char>(hexValue));
    }

    // Close the files
    inputFile.close();
    outputFile.close();

    std::cout << "File decoded successfully." << std::endl;

    return EXIT_SUCCESS;
}

File: HexToFile.java
import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.util.Scanner;

public class HexToFileConverter {

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        // Input hex file path
        System.out.print("Enter the hex file path: ");
        String hexFilePath = scanner.nextLine();

        // Input output file path
        System.out.print("Enter the output file path: ");
        String outputFilePath = scanner.nextLine();

        FileInputStream hexFile = null;
        FileOutputStream outputFile = null;

        try {
            // Open the hex file for reading
            hexFile = new FileInputStream(new File(hexFilePath));

            // Open the output file for writing
            outputFile = new FileOutputStream(new File(outputFilePath));

            // Read hex file and convert to binary
            int byteRead;
            while ((byteRead = hexFile.read()) != -1) {
                // Convert the byte to a hexadecimal string
                String hexByte = String.format("%02x", byteRead & 0xFF);
                // Parse the hexadecimal string to an integer
                int byteValue = Integer.parseInt(hexByte, 16);
                // Write the byte to the output file
                outputFile.write(byteValue);
            }

            System.out.println("File decoded successfully.");
        } catch (IOException e) {
            System.err.println("Error: " + e.getMessage());
        } finally {
            // Close files
            try {
                if (hexFile != null) hexFile.close();
                if (outputFile != null) outputFile.close();
            } catch (IOException e) {
                System.err.println("Error closing files: " + e.getMessage());
            }
        }
    }
}